src/librustc_mir_build/hair/pattern/const_to_pat.rs

```diff
@@ -257,7 +238,92 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
+                            (alloc, offset)
+                        }
+                        _ => unimplemented!(),
+                    };
+                    // The following code is essentially a `deref` operation for slice wide
+                    // pointers to an array. So basically
+                    // ```rust
+                    // const FOO: [T; _] = *SLICE;
+                    // ```
+                    // We can do this because we're the compiler and can read the slice length.
+                    // First, read a pointer.
+                    // FIXME: merge this into `const_deref`, it's much easier to do all this there.
+                    let offset = Pointer::new(AllocId(0), offset);
+                    let ptr = alloc
+                        .read_ptr_sized(&tcx, offset)
+                        .unwrap()
+                        .not_undef()
+                        .unwrap()
+                        .assert_ptr();
+                    // Then read a length.
+                    let offset = offset.offset(tcx.data_layout.pointer_size, &tcx).unwrap();
+                    let len = alloc
+                        .read_ptr_sized(&tcx, offset)
+                        .unwrap()
+                        .not_undef()
+                        .unwrap()
+                        .assert_bits(tcx.data_layout.pointer_size)
+                        .try_into()
+                        .unwrap();
+                    // Now load the memory where the pointer points to.
+                    let data = tcx.alloc_map.lock().unwrap_memory(ptr.alloc_id);
+                    // Now convert the memory into an array with known size.
+                    let array_ty = tcx.mk_array(elem_ty, len);
+                    let cv = tcx.mk_const(ty::Const {
+                        ty: array_ty,
+                        val: ConstKind::Value(ConstValue::ByRef {
+                            alloc: data,
+                            offset: ptr.offset,
+                        }),
+                    });
+                    let prefix = tcx
+                        .destructure_const(param_env.and(cv))
+                        .fields
+                        .iter()
+                        .map(|val| self.recur(val))
+                        .collect();
+                    PatKind::Deref {
+                        subpattern: Pat {
+                            span,
+                            ty: pointee_ty,
+                            kind: Box::new(PatKind::Slice { prefix, slice: None, suffix: vec![] }),
+                        },
+                    }
+                }
+                _ => unimplemented!("{}", cv.ty),
+            },
+            _ => {
+                assert!(
+                    cv.ty.is_primitive() || cv.ty.is_fn_ptr() || cv.ty.is_unsafe_ptr(),
+                    "cannot make a pattern out of {}",
+                    cv.ty
+                );
```

Comment:
Can you whitelist explicitly `ty::Bool | ty::Char | ty::Int(_) | ty::Uint(_) | ty::FnDef(..)` (instead of relying on the methods, which IMO we should probably remove) and make it a proper error?
We shouldn't rely on `type_marked_structural` for this IMO, although that should *also* have a whitelist.

For now you'll have to also allow `ty::RawPtr(_)` and `ty::FnPtr(..)`, just so we can get this PR merged.